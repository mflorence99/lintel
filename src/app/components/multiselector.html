<ng-container
  [formGroup]="multiSelectorForm">

  <div 
    [ngStyle]="{'grid-template-columns': 'repeat(auto-fit, ' + columnWidth + ')'}"
    class="checkboxes">

    <lintel-checkbox
      (mouseout)="description = null"
      (mouseover)="description = getOptionDescription(ix)"
      [formControl]="control"
      [label]="getOptionDecoded(ix)"
      *ngFor="let control of controls; let ix = index">
    </lintel-checkbox>

  </div>

</ng-container>
