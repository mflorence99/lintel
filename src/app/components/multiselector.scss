:host {
  display: block;
}

.multiselector {
  align-items: center;
  display: grid;
  grid-row-gap: 0.5rem;
  grid-template-areas:
    "prefix placeholder suffix"
    "rule   rule        rule"
    ".      checkboxes  checkboxes"
    ".      readout     readout";
  grid-template-columns: auto 1fr auto;
  grid-template-rows: auto 1fr;

  .checkboxes {
    display: grid;
    grid-area: checkboxes;
    grid-template-columns: auto;
    overflow: auto;
  }

  .mat-form-field-prefix {
    grid-area: prefix;
  }

  .mat-form-field-suffix {
    grid-area: suffix;
  }

  .placeholder {
    grid-area: placeholder;
  }

  .rule {
    background-color: var(--vscode-editorWidget-foreground);
    grid-area: rule;
    height: 1px;
  }

  .readout {
    font-size: 0.66rem;
    grid-area: readout;
  }

}
