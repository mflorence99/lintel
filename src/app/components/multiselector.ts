import { ChangeDetectionStrategy } from '@angular/core';
import { ChangeDetectorRef } from '@angular/core';
import { Component } from '@angular/core';
import { ControlValueAccessor } from '@angular/forms';
import { DestroyService } from '../services/destroy';
import { FormArray } from '@angular/forms';
import { FormBuilder } from '@angular/forms';
import { FormControl } from '@angular/forms';
import { FormGroup } from '@angular/forms';
import { Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { OnInit } from '@angular/core';

import { filter } from 'rxjs/operators';
import { forwardRef } from '@angular/core';
import { takeUntil } from 'rxjs/operators';

// NOTE: options can be one of the following:
// -- array of encoded values
// -- array of [encoded, decoded, description] values <-- PREFERRED
// -- array of {
//               [nameOfEncoded]: encoded, 
//               [nameOfDecoded]: decoded, 
//               [nameOfDescription]: description
//             } values

export type MultiselectorOptions = string[] | string[][] | any[];

// NOTE: values can be one of the following:
// -- array of encoded values 
// -- Record<string, boolean> <-- PREFERRED

export type MultiselectorValues = string[] | Record<string, boolean>;

/**
 * Multiselect values via checkboxes
 * 
 * NOTE: just enough to be able to match VSCode as well as possible
 *
 * @see https://blog.thoughtram.io/angular/2016/07/27/custom-form-controls-in-angular-2.html
 *
 * Reference is old, but still helpful
 */

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => MultiselectorComponent),
      multi: true
    },
    DestroyService
  ],
  selector: 'lintel-multiselector',
  templateUrl: 'multiselector.html',
  styleUrls: ['multiselector.scss']
})

export class MultiselectorComponent implements ControlValueAccessor, OnInit {

  @Input() columnWidth = '10rem';

  controls: FormControl[] = [];

  multiSelectorForm: FormGroup;

  @Input() nameOfDecoded = 'value';
  @Input() nameOfDescription = 'description';
  @Input() nameOfEncoded = 'id';

  values = new Set<string>();

  @Input()
  get options(): MultiselectorOptions {
    return this._origOptions;
  }
  set options(options: MultiselectorOptions) {
    if (options && !this._origOptions) {
      this._origOptions = options;
      this._options = this.fromMultiselectorOptions(options);
      this.underConstruction = true;
      // create controls for each option
      // NOTE: now options are [encoded, decoded]
      this.controls = this._options
        .map(option => new FormControl(this.values.has(option[0])));
      const checkboxes = this.multiSelectorForm.controls.checkboxes as FormArray;
      this.controls.forEach(control => checkboxes.push(control));
      this.underConstruction = false;
    }
  }

  @Input()
  get value(): MultiselectorValues {
    return this.toMultiselectorValues();
  }
  set value(values: MultiselectorValues) {
    this.values = this.fromMultiselectorValues(values);
    // patch the form to reflect the values
    this.underConstruction = true;
    const checkboxes = this.multiSelectorForm.controls.checkboxes as FormArray;
    const patch = this._options.map(option => this.values.has(option[0]));
    checkboxes.patchValue(patch, { emitEvent: false });
    this.underConstruction = false;
    this.onChange?.(this.value);
    // TODO: Angular can be so weird!
    this.cdf.detectChanges();
  }

  // these shadow visible properties
  private _options: string[][] = [];
  private _origOptions: MultiselectorOptions;

  private onChange: Function;
  private underConstruction: boolean;
  private valuesType: 'array' | 'object';

  /** ctor  */
  constructor(private cdf: ChangeDetectorRef,
              private destroy$: DestroyService,
              private formBuilder: FormBuilder) {
    this.multiSelectorForm = this.formBuilder.group({
      checkboxes: new FormArray([])
    });
  }

  /** Get an indexed decoded option value */
  getOptionDecoded(ix: number): string {
    return this._options[ix][1];
  }

  /** Get an indexed option description */
  getOptionDescription(ix: number): string {
    return this._options[ix][2];
  }

  /** Get an indexed encoded option value */
  getOptionEncoded(ix: number): string {
    return this._options[ix][0];
  }

  /** When we're ready */
  ngOnInit(): void {
    const checkboxes = this.multiSelectorForm.controls.checkboxes as FormArray;
    checkboxes.valueChanges
      .pipe(
        filter(_ => !this.underConstruction),
        takeUntil(this.destroy$)
      )
      .subscribe((settings: boolean[]) => {
        // NOTE: remember options are [encoded, decoded]
        const values = this._options
          .filter((option, ix) => settings[ix])
          .map(option => option[0]);
        this.values = new Set(values);
        this.onChange?.(this.value);
      });
  }

  /** @see ControlValueAccessor */
  registerOnChange(fn): void {
    this.onChange = fn;
  }

  /** @see ControlValueAccessor */
  registerOnTouched(_): void { }

  /** @see ControlValueAccessor */
  writeValue(value): void {
    this.value = value;
  }

  // private methods

  private fromMultiselectorOptions(options: MultiselectorOptions): string[][] {
    let normalized: string[][] = [];
    // NOTE: see above for different options for supplying options
    if (Array.isArray(options) && (options.length > 0)) {
      if (typeof options[0] === 'string')
        normalized = (options as string[]).map(option => [option, option, null]);
      else if (Array.isArray(options[0]))
        normalized = options as string[][];
      else if (typeof options[0] === 'object')
        normalized = (options as string[]).map(option => [option[this.nameOfEncoded], option[this.nameOfDecoded], option[this.nameOfDescription]]);
    }
    return normalized;
  }

  private fromMultiselectorValues(values: MultiselectorValues): Set<string> {
    if (Array.isArray(values)) {
      this.valuesType = 'array';
      return new Set(values);
    } else if (values && typeof values === 'object') {
      this.valuesType = 'object';
      return new Set(Object.keys(values).filter(val => values[val]));
    } else {
      this.valuesType = 'object';
      return new Set();
    }
  }

  private toMultiselectorValues(): MultiselectorValues {
    if (this.valuesType === 'object') {
      const obj = this._options.reduce((acc, option) => {
        acc[option[0]] = false;
        return acc;
      }, { });
      this.values.forEach(value => obj[value] = true);
      return obj;
    } else return Array.from(this.values) as MultiselectorValues;
  }

}
