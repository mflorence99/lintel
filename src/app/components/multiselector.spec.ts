import { MultiselectorComponent } from './multiselector';
import { TestBed } from '@angular/core/testing';

import { async } from '@angular/core/testing';
import { prepare } from './component.spec';

describe('MultiselectorComponent', () => {

  beforeEach(async(() => prepare()));

  test('Component is created', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });

  test('registerOnChange', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    component.registerOnChange(jest.fn());
    expect(component['onChange']).toBeTruthy();
  });

  test('registerOnTouched', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    expect(component.registerOnTouched(null)).toBeFalsy();
  });

  test('writeValue of array', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    component.writeValue(['london', 'paris', 'paris', 'london']);
    expect(component.value).toEqual(['london', 'paris']);
  });

  test('writeValue of object', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    component.writeValue({ london: true, paris: false, rome: true});
    expect(component.value).toEqual({ london: true, rome: true});
  });

  test('test options as encoded array', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    const options = ['london', 'paris', 'rome'];
    component.options = options;
    expect(component.options).toEqual(options);
    expect(component.getOptionDecoded(1)).toBe('paris');
    expect(component.getOptionEncoded(1)).toBe('paris');
    expect(component.getOptionDescription(1)).toBeFalsy();
  });

  test('test options as encoded/decoded array', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    component.options = [
      ['london', 'London'], 
      ['paris', 'Paris', 'City of Light'], 
      ['rome', 'Rome']
    ];
    expect(component.getOptionDecoded(1)).toBe('Paris');
    expect(component.getOptionEncoded(1)).toBe('paris');
    expect(component.getOptionDescription(1)).toBe('City of Light');
    expect(component.getOptionDescription(2)).toBeFalsy();
  });

  test('test options as array of objects', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    component.options = [
      { id: 'london', value: 'London' },
      { id: 'paris', value: 'Paris', description: 'City of Light' },
      { id: 'rome', value: 'Rome' }
    ];
    expect(component.getOptionDecoded(1)).toBe('Paris');
    expect(component.getOptionEncoded(1)).toBe('paris');
    expect(component.getOptionDescription(1)).toBe('City of Light');
    expect(component.getOptionDescription(2)).toBeFalsy();
  });

  test('a null value is interpreted as all values false', () => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    component.options = ['london', 'paris', 'rome'];
    component.value = null;
    expect(component.value).toEqual({
      london: false,
      paris: false,
      rome: false
    });
  });

  test('ngOnInit', done => {
    const fixture = TestBed.createComponent(MultiselectorComponent);
    const component = fixture.componentInstance;
    component.options = ['london', 'paris', 'rome'];
    const value = { london: true, paris: false, rome: true };
    component.value = value;
    component.registerOnChange(value => {
      expect(value).toEqual(value);
      done();
    });
    component.ngOnInit();
    // NOTE: trips valueChanges
    component.multiSelectorForm.controls.checkboxes.updateValueAndValidity({ emitEvent: true });
  });

});
