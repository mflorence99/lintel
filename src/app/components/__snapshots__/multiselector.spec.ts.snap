// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiselectorComponent snapshot 1`] = `
<lintel-multiselector
  __ngContext__={[Function LRootView]}
  _options={[Function Array]}
  cdf={[Function ViewRef]}
  columnWidth={[Function String]}
  controls={[Function Array]}
  destroy$={[Function DestroyService]}
  enabled={[Function Boolean]}
  formBuilder={[Function FormBuilder]}
  multiSelectorForm={[Function FormGroup]}
  nameOfDecoded={[Function String]}
  nameOfDescription={[Function String]}
  nameOfEncoded={[Function String]}
  origOptions={[Function Array]}
  underConstruction="false"
  values={[Function Set]}
>
  <p
    class="checkboxes"
    style="grid-template-columns: repeat(auto-fit, 10rem);"
  >
    <lintel-checkbox
      class="ng-untouched ng-pristine ng-valid"
      title=""
    >
      <button
        class="checkbox"
        tabindex="0"
      >
        <fa-icon
          class="ng-fa-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check fa-w-16"
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </fa-icon>
      </button>
      <label
        class="label"
      >
         London
  
      </label>
    </lintel-checkbox>
    <lintel-checkbox
      class="ng-untouched ng-pristine ng-valid"
      title="City of Light"
    >
      <button
        class="checkbox"
        tabindex="0"
      >
        <fa-icon
          class="ng-fa-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check fa-w-16"
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </fa-icon>
      </button>
      <label
        class="label"
      >
         Paris
  
      </label>
    </lintel-checkbox>
    <lintel-checkbox
      class="ng-untouched ng-pristine ng-valid"
      title=""
    >
      <button
        class="checkbox"
        tabindex="0"
      >
        <fa-icon
          class="ng-fa-icon"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check fa-w-16"
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
        </fa-icon>
      </button>
      <label
        class="label"
      >
         Rome
  
      </label>
    </lintel-checkbox>
    
  </p>
</lintel-multiselector>
`;
